### ISSUE TITLE
sh.exe crashes when an environment variable called "ö" exists

### ISSUE THREAD

**Issue by albinsunnanbo** [git-for-windows/git#1137]
- [ X ] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
$ git --version --build-options

git version 2.12.2.windows.2
built from commit: 7723f38cfb0e80f010afaebdd9fec4d0822fd2e1
sizeof-long: 4
machine: x86_64
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.14393]

Win 10 64 bit
Had the same issue on Win 8 64 bit
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

Defaults

```
# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"
> type "C:\Program Files (x86)\Git\etc\install-options.txt"
> type "%USERPROFILE%\AppData\Local\Programs\Git\etc\install-options.txt"
$ cat /etc/install-options.txt

Path Option: Cmd
Plink Path: C:\Program Files\TortoiseGit\bin\TortoiseGitPlink.exe
SSH Option: Plink
CURL Option: OpenSSL
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Performance Tweaks FSCache: Enabled
Use Credential Manager: Enabled
Enable Symlinks: Disabled
```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

not that I can think of

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

cmd

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

![image](https://cloud.githubusercontent.com/assets/5158396/25026745/619ab8da-20a8-11e7-89a1-b4827301125d.png)
```
In clear text:
-----------------

Microsoft Windows [Version 10.0.14393]
(c) 2016 Microsoft Corporation. All rights reserved.

C:\Users\[username]>"\Program Files\Git\bin\sh.exe"

[username]@[computername] MINGW64 ~
$ git --version
git version 2.12.2.windows.2

[username]@[computername] MINGW64 ~
$ exit

C:\Users\[username]>set ö=1

C:\Users\[username]>"\Program Files\Git\bin\sh.exe"
      1 [main] bash 4768 C:\Program Files\Git\bin\..\usr\bin\bash.exe: *** fatal error - internal error reading the windows environment - too many environment variables?
    709 [main] bash 4768 cygwin_exception::open_stackdumpfile: Dumping stack trace to bash.exe.stackdump

C:\Users\[username]>

```
 - What did you expect to occur after running these commands?

I expected sh.exe/bash.exe to not crash when there was an environment variable called "ö"

 - What actually happened instead?

sh.exe crashed with the stack trace above.

 - If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?

Doesn't even have to be in a repository.

Note: This affects a lot of git functionality. Including rebase from Git Extensions.

**Comment by dscho** [comment]
This is quite possibly a bug in [Cygwin](https://cygwin.com). Can you verify my hunch by installing it and launching a Cygwin Bash?

**Comment by dscho** [comment]
Thanks for your help...

I verified that this is indeed a problem with Cygwin:

```cmd
D:\cygwin64>set ö=1

D:\cygwin64>Cygwin.bat
      0 [main] bash 27176 D:\cygwin64\bin\bash.exe: *** fatal error - internal error reading the windows environment - too many environment variables?
    411 [main] bash 27176 cygwin_exception::open_stackdumpfile: Dumping stack trace to bash.exe.stackdump
```

If this problem is serious enough to prevent you from working, you may want to consider debugging it by [building the MSYS2 runtime yourself](https://github.com/git-for-windows/git/wiki/Building-msys2-runtime#rebuild-the-msys2-runtime) possibly adding [some debug statements](https://github.com/git-for-windows/msys2-runtime/commit/2dfd6ad4d89d13df13eeda720ee4269e279029cd). Alternatively, you can try your luck with the Cygwin maintainers: https://cygwin.com/problems.html

**Comment by dscho** [comment]
Seems that the original reporter took my suggestion to heart and reported this, and that the excellent Cygwin developers fixed this issue:

```
 - Fix crash if env var name starts with non-ASCII char
  Addresses: https://cygwin.com/ml/cygwin/2017-05/msg00330.html
```

It would appear that the tireless Cygwin developers fixed this in cygwin/cygwin@7afc1124b6530adc07b63ba691725311fada4012
